### PR TITLE
First minimal version with Meta-Data POST/GET

### PR DESCRIPTION
**Pull request by jsiloto**
This was entirely built by modifying dojot/device-manager and some leftover code may still be around.
Also, the PR in https://github.com/dojot/docker-compose/pull/29 is needed for testing

### CODE REVIEW

File: ImageManager/ImageManager.py [hunk 1]
```diff
@@ -0,0 +1,89 @@
+"""
+    Handles CRUD operations for images, and their configuration on the
+    FIWARE backend
```
Comment (mmagr) [inline ImageManager/ImageManager.py]: we're not configuring fiware, are we?

File: ImageManager/ImageManager.py [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+global_imageid = 20
+
+
+@image.route('/image/<imageid>', methods=['GET'])
+def get_image(imageid):
+    try:
+        print(request)
+        init_tenant_context(request, db)
+        orm_image = assert_image_exists(imageid)
+        result = {
```
Comment (mmagr) [inline ImageManager/ImageManager.py]: why not have this include creation/update dates as well?
and if doing so, why not have marshmallow do this for us?

File: ImageManager/ImageManager.py [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+LOGGER.addHandler(logging.StreamHandler())
+LOGGER.setLevel(logging.INFO)
+
+global_imageid = 20
+
+
+@image.route('/image/<imageid>', methods=['GET'])
+def get_image(imageid):
+    try:
+        print(request)
```
Comment (mmagr) [inline ImageManager/ImageManager.py]: I think it'd be better to make use of LOGGER for log generation.
If this is just left over from development, we could remove it.

File: ImageManager/DatabaseModels.py [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+from flask_sqlalchemy import SQLAlchemy
+from app import app
+from utils import HTTPRequestError
+from conf import CONFIG
+from sqlalchemy.exc import IntegrityError
+
+app.config['SQLALCHEMY_DATABASE_URI'] = CONFIG.get_db_url()
+app.config['SQLALCHEMY_TRACK_MODIFICATIONS'] = False
+
+print(app.config['SQLALCHEMY_DATABASE_URI'])
```
Comment (mmagr) [inline ImageManager/DatabaseModels.py]: This should be removed, shouldn't it?

File: ImageManager/ImageManager.py [hunk 5]
```diff
@@ -0,0 +1,89 @@
...
+        # imageid = generate_image_id()
+        global global_imageid
+        imageid = str(global_imageid)
+        global_imageid = global_imageid + 1
+        image_data['id'] = imageid
+
+        orm_image = Image(**image_data)
+        db.session.add(orm_image)
+
+        # # TODO remove this in favor of kafka as data broker....
```
Comment (mmagr) [inline ImageManager/ImageManager.py]: I think those can be safely removed from code - this service should require no events of that nature being sent at this time.

File: ImageManager/ImageManager.py [hunk 6]
```diff
@@ -0,0 +1,89 @@
...
+
+        try:
+            db.session.commit()
+        except IntegrityError as error:
+            handle_consistency_exception(error)
+
+        else:
+            result = {'message': 'image updated', 'image': imageid}
+
+        print("Operation Successful - result: {}".format(result))
```
Comment (mmagr) [inline ImageManager/ImageManager.py]: the same note on logging here

File: ImageManager/conf.py [hunk 7]
```diff
@@ -0,0 +1,35 @@
...
+                 kafka_host="kafka",
+                 kafka_port="9092",
+                 create_db=True):
+        self.dbname = os.environ.get('DBNAME', db)
+        self.dbhost = os.environ.get('DBHOST', dbhost)
+        self.dbuser = os.environ.get('DBUSER', dbuser)
+        self.dbpass = os.environ.get('DBPASS', dbpass)
+        self.dbdriver = os.environ.get('DBDRIVER', dbdriver)
+        self.create_db = os.environ.get('CREATE_DB', create_db)
+        self.kafka_host = os.environ.get('KAFKA_HOST', kafka_host)
```
Comment (mmagr) [inline ImageManager/conf.py]: we might want to remove kafka config

File: ImageManager/conf.py [hunk 8]
```diff
@@ -0,0 +1,35 @@
...
+
+class Config(object):
+    """ Abstracts configuration, either retrieved from environment or from ctor arguments """
+    def __init__(self,
+                 db="dojot_imgm",
+                 dbhost="postgres",
+                 dbuser="postgres",
+                 dbpass=None,
+                 dbdriver="postgresql+psycopg2",
+                 kafka_host="kafka",
```
Comment (mmagr) [inline ImageManager/conf.py]: we might want to remove kafka config

File: ImageManager/utils.py [hunk 9]
```diff
@@ -0,0 +1,53 @@
...
+        payload = json.dumps({'message': message, 'status': status})
+    elif 200 <= status < 300:
+        payload = json.dumps({'message': 'ok', 'status': status})
+    else:
+        payload = json.dumps({'message': 'Request failed', 'status': status})
+
+    return make_response(payload, status)
+
+
+def create_id():
```
Comment (mmagr) [inline ImageManager/utils.py]: This could be removed

File: LICENSE [hunk 10]
```diff
@@ -47,7 +47,7 @@ authors' sake, the GPL requires that modified versions be marked as
 changed, so that their problems will not be attributed erroneously to
 authors of previous versions.
 
-  Some devices are designed to deny users access to install or run
+  Some images are designed to deny users access to install or run
```
Comment (mmagr) [inline LICENSE]: this must be rolled back (we shouldn't tinker with the license). Perhaps this is the result of a "replace all"?

File: setup.py [hunk 11]
```diff
@@ -0,0 +1,49 @@
...
+    name='ImageManager',
+    description='Dojot image manager.',
+    version=':versiontools:ImageManager:',
+
+    packages=setuptools.find_packages(),
+    include_package_data=True,
+    install_requires=get_requirements('requirements/requirements.txt'),
+    setup_requires=('versiontools'),
+
+    author='Matheus Magalhaes',
```
Comment (mmagr) [inline setup.py]: hmm, this should be Juliano Siloto, right? :thinking:

File: tests/autotest.sh [hunk 12]
```diff
@@ -0,0 +1,215 @@
+#!/bin/bash
```
Comment (mmagr) [inline tests/autotest.sh]: This file should be removed. I don't suppose it works even for device manager by now.
(furthermore, if anyone is to do this sort of automatic testing, it's Travis, not Jenkins)

File: tests/docker-compose.yaml [hunk 13]
```diff
@@ -0,0 +1,26 @@
+version: '2'
```
Comment (mmagr) [inline tests/docker-compose.yaml]: Same as for `tests/autotest.sh` goes here.

**Review by mmagr — CHANGES_REQUESTED** [review]
As for testing, in light of the troubles we had when setting up the dev environment, a PR dojot/device-manager#21 has been submitted.

Basically what it does is to create a local compose file allowing for easier testing of the isolated service. I think one might benefit from that here too.

**Comment by jsiloto** [comment]
All suggested changes were made except for the local docker-compose addition.
The PR dojot/device-manager#21 is enought to get the service up and running but the service is inacessible from outside world

**Review by mmagr — APPROVED** [review]

**Comment by jsiloto** [comment]
My bad, the previous reply was wrong due my poor understanding of docker. But since the whole point of how-to's is for people who don't know the inner workings of the system, I've opened an issue related to this: dojot/image-manager/issues/3

**Pull request merged**